### ISSUE TITLE
inf entries in spec file

### ISSUE THREAD

**Issue by jhmatthews** [sirocco-rt/sirocco#95]
I've recently discovered slightly worrying behaviour in some spec output files. I've noticed that this has happened in a few different types of macro atom CV models on Iridis -- to reproduce, run https://github.com/agnwinds/python/commit/514241705a872f72eab65714205981acbfeb5b35 on ~/Dropbox/Python/bug_pf_files/cv_alpha4_r4r12_mdot1e9_rv7e10_opt.pf 

Some, seemingly slightly random, columns in the extracted spectrum show up as inf (see below image):

![screenshot 2014-07-30 10 38 40](https://cloud.githubusercontent.com/assets/3007249/3747341/52146f28-17cd-11e4-84e4-1ec44c13e7f0.png)

I've not had a chance to look into this, but it would seem that obvious things for me to check are:
- is the problem macro atom specific?
- is this related to the stellar atmospheres in any way?
- is this related to extract (being as the total summed columns seem to be OK)?
- does the problem occur in Python 78 (i.e. is it a recent change which causes this? -- I am suspecting so, as I hadn't noticed this before.)

I've assigned this to myself and flagged as urgent. It will definitely need to be fixed before CV paper as we will want to present runs such as this.

**Comment by kslong** [comment]
Hi James,
It might be worthwhile to see whether this problem has appeared recently.  I do not recall ever seeing this before.
Knox

**Comment by jhmatthews** [comment]
I'm wondering if this is related to #83 , normalising the anisotropic scattering, because we now do the following in trans_phot.c before extracting the photon

``` C
p_norm = NNSCAT_SAFETY * (1. - exp (-tau_norm)) / tau_norm;
pextract.w *= p[nphot].nnscat / p_norm;
```

So perhaps if tau_norm is high enough we can get zero here and hence an infinite weight. I'm testing if this is the problem. If it is I'll add a condition which sets p_norm to a minimum value in randwind_thermal_trapping(), and also an Error and sane_check in trans_phot to check weights are in bounds.

**Comment by jhmatthews** [comment]
Ok, figured this one out. It's similar to my suspicions above, in that it's p_norm which is responsible, but it's actually because of how the function behaves for **low** tau which is the problem. If you have a low value of tau (<1e-16 or so) then the function 1-exp(-tau) goes to zero...you can see how the function P_Escape behaves here:

![tau](https://cloud.githubusercontent.com/assets/3007249/3856365/b83a33a8-1ef5-11e4-8f91-c4da08076ccd.png)

I should have spotted this because it's the same reason we but lines like the following in the p_escape function:

``` C
      if (tau < 1e-6)
    escape = 1.;
      else if (tau < 10.0)
    escape = (1. - exp (-tau)) / tau;
      else
    escape = 1. / tau;
```

The fix is simple- simply add if statements like the following to trans_phot.c and anisowind.c. My mistake!

**Comment by jhmatthews** [comment]
To stop something like this happening in future, I think it would be good to have a function p_escape_from_tau(tau) which just does the following

``` C
#define TAU_MIN 1e-6
double
p_escape_from_tau(tau)
double tau;
{
  double x;

  if (tau < TAU_MIN)
    escape = 1.;
  else if (tau < 10.0)
    escape = (1. - exp (-tau)) / tau;
  else
    escape = 1. / tau;

  return (escape);
}
```

this takes care of the numerical issues in the function itself and ensures that this is done whenever it is called.

**Comment by jhmatthews** [comment]
Closed by pull request #98
